diff --git a/os/hal/ports/STM32/STM32L1xx/platform.mk b/os/hal/ports/STM32/STM32L1xx/platform.mk
index e2c2344af..dcf6f83bc 100644
--- a/os/hal/ports/STM32/STM32L1xx/platform.mk
+++ b/os/hal/ports/STM32/STM32L1xx/platform.mk
@@ -17,8 +17,8 @@ ifneq ($(findstring HAL_USE_ADC TRUE,$(HALCONF)),)
 PLATFORMSRC += $(CHIBIOS)/os/hal/ports/STM32/STM32L1xx/hal_adc_lld.c
 endif
 else
-               $(CHIBIOS)/os/hal/ports/STM32/STM32L1xx/hal_adc_lld.c
-PLATFORMSRC += $(CHIBIOS)/os/hal/ports/STM32/STM32L1xx/hal_ext_lld_isr.c
+PLATFORMSRC += $(CHIBIOS)/os/hal/ports/STM32/STM32L1xx/hal_adc_lld.c \
+               $(CHIBIOS)/os/hal/ports/STM32/STM32L1xx/hal_ext_lld_isr.c
 endif
 
 # Drivers compatible with the platform.
